### PR TITLE
Custom Resolution 2/?

### PR DESCRIPTION
**Pull request by BlisterB**
#2873

The purpose of this PR is to adapt all elements of battles with custom resolutions;

- [x] Battle 2000
     - [X] UI
     - [x] Animation
     - [x] Background
     - [x] Monster
- [x] Battle 2003
     - [X] UI
     - [x] Animation
     - [x] Background
     - [X] Monster
     - [X] Heroes
     - [x] Cursor
     - [x] Texts

- [x] Allow reading resolution from INI files

### CODE REVIEW

**Comment by BlisterB** [comment]
The UI of battles have been centered for 2K and 2K3.
All configurations (2k and the 6 variants of 2k3) have been tested, but it will probably need more tests at the end.
image
image

When the UI displays moving boxes with a widescreen resolution, it's doesn't feel right (we see boxes that should be outside of the screen). It's maybe possible to hide boxes that should be outside of a 3:4 screen.
image

I also changed the way Message Box works.
After my previous PR, Message Box used to be displayed from 3% of border if height > 240px.
It created some edge cases (ex. if height is 241px) and I discovered that it created a visual bug at the end of 2K battles.
image


I decided to harmonize MessageBox with all other systems, ie. displaying the UI in a centered 320x240px frame. Messages are now slightly more distant from borders with 480x272px resolution.
image

I also harmonized the gold windows which didn't comply with this policy.
image

(NB: The Load/Save menu doesn't comply with this policy because of the vertical nature of this system)

**Comment by BlisterB** [comment]
Battlers are now centered in 2k3 battles 

Before:
image

After:
image

**Comment by BlisterB** [comment]
I noticed that on map, battle animations works correctly when targetting an event. However when they are shown on the entire map, they loop horizontally and vertically.
image

In battle, they don't target correctly enemies.

**Comment by Ghabry** [comment]
That battle effects Loop when targeting the screen on the map is normal

**Comment by BlisterB** [comment]
@Ghabry Oh you are right, It's the same thing with RPG_RT.
I'm surprised of this behavior. oO

**Comment by BlisterB** [comment]
image
I tried this code, but it doesn't work because dst.GetWidth() and dst.GetHeight() returns the size of the screen (not 320x240px).
I tried to understand where this dst was created, but with no luck.
Is there a way to retrieve the size of the image in order to compare it with the screen target size ?
The idea : if the size of the image is shorter than the size of the screen target: we center it.

**Comment by Ghabry** [comment]
Draw gets the render target passed. Usually the screen.

You want to check fg_bitmap or bg_bitmap. Can't remember which one it is. Note that 2k3 supports two backgrounds at once.

**Comment by BlisterB** [comment]
Thank you @Ghabry 😁

The background is now centered when it doesn't fill the screen.

**Comment by BlisterB** [comment]
It should be good for battle 👍
I may have miss some things to center (I suspect some call of `GetBattlePosition()` in `scene_battle_rpg2k3.cpp`), this PR would need testing to reveal anomalies imo.

**Comment by BlisterB** [comment]
Title screen is now center when it doesn't fill the screen
image

**Comment by BlisterB** [comment]
The resolution of the Player now changes when necessary.
- When starting on GameBrowser, it uses the base resolution of 320x240.
- When starting a game, it reads WinW/WinH value in RPG_RT.ini and adapt the resolution.
- When exiting a game, it return to 320x240 in the GameBrowser.


@Ghabry Thank you for the code, unfortunatly the windows doesn't resize after changing the resolution :(.

**Comment by Ghabry** [comment]
@BlisterB 

There is just a lot of ``TARGET_WIDHT/HEIGHT`` stuff hardcoded in SDL2.

By changing some stuff is appears to work though the Zoom key is broken for unknown reasons :sweat_smile: .

```diff
diff --git a/src/platform/sdl/sdl2_ui.cpp b/src/platform/sdl/sdl2_ui.cpp
index 70e19305..c0da172c 100644
--- a/src/platform/sdl/sdl2_ui.cpp
+++ b/src/platform/sdl/sdl2_ui.cpp
@@ -218,6 +218,8 @@ bool Sdl2Ui::ChangeDisplaySurfaceResolution(int new_width, int new_height) {
 
 	main_surface = new_main_surface;
 
+	window.size_changed = true;
+
 	return true;
 }
 
@@ -446,7 +448,7 @@ void Sdl2Ui::ToggleZoom() {
 	// get current window size, calculate next bigger zoom factor
 	int w, h;
 	SDL_GetWindowSize(sdl_window, &w, &h);
-	last_display_mode.zoom = std::min(w / SCREEN_TARGET_WIDTH, h / SCREEN_TARGET_HEIGHT);
+	last_display_mode.zoom = std::min(w / main_surface->width(), h / main_surface->height());
 	current_display_mode.zoom = last_display_mode.zoom + 1;
 
 	// get maximum usable window size
@@ -456,8 +458,8 @@ void Sdl2Ui::ToggleZoom() {
 	SDL_GetDisplayUsableBounds(display_index, &max_mode);
 
 	// reset zoom, if it does not fit
-	if ((max_mode.h < SCREEN_TARGET_HEIGHT * current_display_mode.zoom) ||
-		(max_mode.w < SCREEN_TARGET_WIDTH * current_display_mode.zoom)) {
+	if ((max_mode.h < main_surface->height() * current_display_mode.zoom) ||
+		(max_mode.w < main_surface->width() * current_display_mode.zoom)) {
 		current_display_mode.zoom = 1;
 	}
 	EndDisplayModeChange();
@@ -496,20 +498,20 @@ void Sdl2Ui::UpdateDisplay() {
 		float width_float = static_cast<float>(window.width);
 		float height_float = static_cast<float>(window.height);
 
-		constexpr float want_aspect = (float)SCREEN_TARGET_WIDTH / SCREEN_TARGET_HEIGHT;
+		float want_aspect = (float)main_surface->width() / main_surface->height();
 		float real_aspect = width_float / height_float;
 
 		if (scaling_mode == ScalingMode::Integer) {
 			// Integer division on purpose
 			if (want_aspect > real_aspect) {
-				window.scale = static_cast<float>(window.width / SCREEN_TARGET_WIDTH);
+				window.scale = static_cast<float>(window.width / main_surface->width());
 			} else {
-				window.scale = static_cast<float>(window.height / SCREEN_TARGET_HEIGHT);
+				window.scale = static_cast<float>(window.height / main_surface->height());
 			}
 
-			viewport.w = static_cast<int>(ceilf(SCREEN_TARGET_WIDTH * window.scale));
+			viewport.w = static_cast<int>(ceilf(main_surface->width() * window.scale));
 			viewport.x = (window.width - viewport.w) / 2;
-			viewport.h = static_cast<int>(ceilf(SCREEN_TARGET_HEIGHT * window.scale));
+			viewport.h = static_cast<int>(ceilf(main_surface->height() * window.scale));
 			viewport.y = (window.height - viewport.h) / 2;
 
 			SDL_RenderSetViewport(sdl_renderer, &viewport);
@@ -519,18 +521,18 @@ void Sdl2Ui::UpdateDisplay() {
 			SDL_RenderSetViewport(sdl_renderer, nullptr);
 		} else if (want_aspect > real_aspect) {
 			// Letterboxing (black bars top and bottom)
-			window.scale = width_float / SCREEN_TARGET_WIDTH;
+			window.scale = width_float / main_surface->width();
 			viewport.x = 0;
 			viewport.w = window.width;
-			viewport.h = static_cast<int>(ceilf(SCREEN_TARGET_HEIGHT * window.scale));
+			viewport.h = static_cast<int>(ceilf(main_surface->height() * window.scale));
 			viewport.y = (window.height - viewport.h) / 2;
 			SDL_RenderSetViewport(sdl_renderer, &viewport);
 		} else {
 			// black bars left and right
-			window.scale = height_float / SCREEN_TARGET_HEIGHT;
+			window.scale = height_float / main_surface->height();
 			viewport.y = 0;
 			viewport.h = window.height;
-			viewport.w = static_cast<int>(ceilf(SCREEN_TARGET_WIDTH * window.scale));
+			viewport.w = static_cast<int>(ceilf(main_surface->width() * window.scale));
 			viewport.x = (window.width - viewport.w) / 2;
 			SDL_RenderSetViewport(sdl_renderer, &viewport);
 		}
@@ -541,7 +543,7 @@ void Sdl2Ui::UpdateDisplay() {
 			}
 			SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "linear");
 			sdl_texture_scaled = SDL_CreateTexture(sdl_renderer, texture_format, SDL_TEXTUREACCESS_TARGET,
-			   static_cast<int>(ceilf(window.scale)) * SCREEN_TARGET_WIDTH, static_cast<int>(ceilf(window.scale)) * SCREEN_TARGET_HEIGHT);
+			   static_cast<int>(ceilf(window.scale)) * main_surface->width(), static_cast<int>(ceilf(window.scale)) * main_surface->height());
 			SDL_SetHint(SDL_HINT_RENDER_SCALE_QUALITY, "nearest");
 			if (!sdl_texture_scaled) {
 				Output::Debug("SDL_CreateTexture failed : {}", SDL_GetError());

```

With this code the results are:

WinW=400, WinH=200

screenshot_1

WinW=400, WinH=400

screenshot_3

WinW=400, WinH=800

screenshot_4

**Comment by Ghabry** [comment]
btw you likely want to rebase your branch on the upstream master branch because the settings scene was added.

Problem is here that this gives a conflict.

Step-by-Step how to resolve it:

**BACKUP YOUR PLAYER FOLDER!!!** Then:

```
# Add upstream repository
git remote add upstream https://github.com/easyrpg/player

# Update upstream repository
git fetch upstream

# Do a rebase on master
git rebase upstream/master --autostash
```

This will give a conflict in player.cpp

- Open player.cpp and search for ``====``

Replace this

```cpp
<<<<<<< HEAD
                DisplayUi = BaseUi::CreateUi(SCREEN_TARGET_WIDTH, SCREEN_TARGET_HEIGHT, cfg);
=======
                DisplayUi = BaseUi::CreateUi(Player::screen_width, Player::screen_height, cfg.video);
>>>>>>> 042cf567 (Custom Resolution : Change the resolution of the Player when necessary)

```

with that

```
DisplayUi = BaseUi::CreateUi(Player::screen_width, Player::screen_height, cfg);
```

- Save player.cpp

```
# Commit the changes
git add src/player.cpp

# Continue the rebase
git rebase --continue

# This usually opens vi. Type ":q!" to exit
```

- The rebase should complete succesfully. If anything failed type ``git rebase --abort``.
- Push the changes: ``git push --force-with-lease origin``

git interface sucks :P

**Comment by BlisterB** [comment]
Hello Ghabry, thank you I followed your instruction successfully. 👍

The switch of resolutions work flawlessly in fullscreen. 👍
In window mode, the window size doesn't change, also we cannot use F5 with a custom resolution (but if we go back to gamebrowser, it works again).

I'll start the migration of SCREEN_TARGET_WIDTH/HEIGHT, I can see the end of the tunnel. 😁

**Review by Ghabry — COMMENTED** [review]
(please do these changes in a seperate commit)

I will also add one/two commits to change some stuff related to SDL2 because, so don't touch this ^^'

**Comment by Ghabry** [comment]
@BlisterB 

- Fixed Zoom
- Added error handling
- Scene settings supported
- Fixed missing resolution switching for some components (MessageOverlay)

To get the changes:

```
# Save local changes
git stash save

# Get remote changes
git fetch
git pull

# Restore local changes
git stash pop
```

If you already have an additional local commit you must ``git rebase origin/master --autostash`` instead, otherwise it creates a merge commit :/.

**Comment by Ghabry** [comment]
Demonstration video: https://easyrpg.org/downloads/tmp/resolution-0.7.1.mp4

Obvious bug is the incorrectly centered text in the "Exit Game?" prompt.

**Comment by BlisterB** [comment]
Thank you it works flawlessly. Wow, we did it!! 😁

**Comment by Ghabry** [comment]
There are multiple instances related to the Panorama where the amount of tiles that fit on the screen are used.

Width (20):

plane.cpp

```cpp
int bg_width = std::max(Game_Map::GetWidth() * TILE_SIZE, 20 * TILE_SIZE);
```

game_map.cpp

```cpp
void Game_Map::Parallax::ResetPositionX() {
	Params params = GetParallaxParams();

	if (params.name.empty()) {
		return;
	}

	if (!params.scroll_horz && !LoopHorizontal()) {
		if (GetWidth() > 20 && parallax_width > Player::screen_width) {
			const int w = (GetWidth() - 20) * TILE_SIZE;
			const int ph = 2 * std::min(w, parallax_width - Player::screen_width) * map_info.position_x / w;
			if (Player::IsRPG2k()) {
				SetPositionX(ph);
			} else {
				// 2k3 does not do the (% parallax_width * TILE_SIZE * 2) here
				panorama.pan_x = ph;
			}
		} else {
			panorama.pan_x = 0;
		}
	}
}
```

Height (15):

```cpp
void Game_Map::Parallax::ResetPositionY() {
	Params params = GetParallaxParams();

	if (params.name.empty()) {
		return;
	}

	if (!params.scroll_vert && !Game_Map::LoopVertical()) {
		if (GetHeight() > 15 && parallax_height > Player::screen_height) {
			const int h = (GetHeight() - 15) * TILE_SIZE;
			const int pv = 2 * std::min(h, parallax_height - Player::screen_height) * map_info.position_y / h;
			SetPositionY(pv);
		} else {
			panorama.pan_y = 0;
		}
	}
}
```

Problem here is that with custom W/H there can be fractional amount of tiles on the screen so this likely also needs floating point math :thinking:

**Comment by Ghabry** [comment]
For reference for a legacy mode later. The absolute minimum to position Picture correctly and to lie about the screen scroll position:

This doesn't fake everything but is a start that should work for alot of games ;)

```diff
diff --git a/src/game_interpreter_control_variables.cpp b/src/game_interpreter_control_variables.cpp
index 4fe1a2274..753431ba8 100644
--- a/src/game_interpreter_control_variables.cpp
+++ b/src/game_interpreter_control_variables.cpp
@@ -28,6 +30,7 @@
 #include "utils.h"
 #include <cmath>
 #include <cstdint>
+#include <lcf/rpg/savepartylocation.h>
 
 int ControlVariables::Random(int value, int value2) {
 	int rmax = std::max(value, value2);
@@ -183,13 +186,18 @@ int ControlVariables::Event(int op, int event_id, const Game_Interpreter& interp
 						dir == 1 ? 6 :
 						dir == 2 ? 2 : 4;
 				break;
-			case 4:
+			case 4: {
 				// Screen X
-				return character->GetScreenX();
-				break;
-			case 5:
+				int pan_delta = (Game_Player::GetDefaultPanX() - lcf::rpg::SavePartyLocation::kPanXDefault) / TILE_SIZE;
+
+				return character->GetScreenX() - pan_delta;
+			}
+			case 5: {
 				// Screen Y
-				return character->GetScreenY();
+				int pan_delta = (Game_Player::GetDefaultPanY() - lcf::rpg::SavePartyLocation::kPanYDefault) / TILE_SIZE;
+
+				return character->GetScreenY() - pan_delta;
+			}
 		}
 
 		Output::Warning("ControlVariables::Event: Unknown op {}", op);
diff --git a/src/game_player.cpp b/src/game_player.cpp
index 45024c70d..aede8f185 100644
--- a/src/game_player.cpp
+++ b/src/game_player.cpp
@@ -43,16 +43,6 @@
 #include <cmath>
 #include "scene_gameover.h"
 
-namespace {
-	int default_pan_x() {
-		return (Utils::RoundTo<int>(static_cast<float>(Player::screen_width) / TILE_SIZE / 2) - 1) * SCREEN_TILE_SIZE;
-	}
-
-	int default_pan_y() {
-		return (Utils::RoundTo<int>(static_cast<float>(Player::screen_height) / TILE_SIZE / 2) - 1) * SCREEN_TILE_SIZE;
-	}
-}
-
 Game_Player::Game_Player(): Game_PlayerBase(Player)
 {
 	SetDirection(lcf::rpg::EventPage::Direction_down);
@@ -155,10 +145,10 @@ void Game_Player::MoveTo(int map_id, int x, int y) {
 
 		// pan_state does not reset when you change maps.
 		data()->pan_speed = lcf::rpg::SavePartyLocation::kPanSpeedDefault;
-		data()->pan_finish_x = default_pan_x();
-		data()->pan_finish_y = default_pan_y();
-		data()->pan_current_x = default_pan_x();
-		data()->pan_current_y = default_pan_y();
+		data()->pan_finish_x = GetDefaultPanX();
+		data()->pan_finish_y = GetDefaultPanY();
+		data()->pan_current_x = GetDefaultPanX();
+		data()->pan_current_y = GetDefaultPanY();
 
 		ResetAnimation();
 
@@ -764,6 +754,14 @@ void Game_Player::SetEncounterSteps(int steps) {
 	data()->encounter_steps = steps;
 }
 
+int Game_Player::GetDefaultPanX() {
+	return (Utils::RoundTo<int>(static_cast<float>(Player::screen_width) / TILE_SIZE / 2) - 1) * SCREEN_TILE_SIZE;
+}
+
+int Game_Player::GetDefaultPanY() {
+	return (Utils::RoundTo<int>(static_cast<float>(Player::screen_height) / TILE_SIZE / 2) - 1) * SCREEN_TILE_SIZE;
+}
+
 void Game_Player::LockPan() {
 	data()->pan_state = lcf::rpg::SavePartyLocation::PanState_fixed;
 }
@@ -793,8 +791,8 @@ void Game_Player::StartPan(int direction, int distance, int speed) {
 }
 
 void Game_Player::ResetPan(int speed) {
-	data()->pan_finish_x = default_pan_x();
-	data()->pan_finish_y = default_pan_y();
+	data()->pan_finish_x = GetDefaultPanX();
+	data()->pan_finish_y = GetDefaultPanY();
 	data()->pan_speed = 2 << speed;
 }
 
diff --git a/src/game_player.h b/src/game_player.h
index 95db5ed8d..9e009ee70 100644
--- a/src/game_player.h
+++ b/src/game_player.h
@@ -134,6 +134,9 @@ public:
 	int GetTargetPanX() const;
 	int GetTargetPanY() const;
 
+	static int GetDefaultPanX();
+	static int GetDefaultPanY();
+
 	void LockPan();
 	void UnlockPan();
 	void StartPan(int direction, int distance, int speed);
diff --git a/src/sprite_picture.cpp b/src/sprite_picture.cpp
index c4b15771c..56a11e374 100644
--- a/src/sprite_picture.cpp
+++ b/src/sprite_picture.cpp
@@ -96,8 +96,8 @@ void Sprite_Picture::Draw(Bitmap& dst) {
 		y -= Main_Data::game_screen->GetShakeOffsetY();
 	}
 
-	SetX(x);
-	SetY(y);
+	SetX(x + Player::menu_offset_x);
+	SetY(y + Player::menu_offset_y);
 	SetZoomX(data.current_magnify / 100.0);
 	SetZoomY(data.current_magnify / 100.0);
 
```

**Comment by Ghabry** [comment]
hm actually "WinW" and "WinH" could work as they do now as they are for games designed for it and old games could have it as a user setting:

screenshot_3

screenshot_4

Not asking you to add this. I'm just brainstorming :laughing:

**Review by Ghabry — APPROVED** [review]
Only needs these changes applied to fix some Panorama cases:
https://github.com/EasyRPG/Player/pull/2897#issuecomment-1448941172

Otherwise looks good to me as an exprimental feature. There will be issues with it later but is a solid base :)

Thanks for this unexpected contribution.

**Comment by Ghabry** [comment]
Patch to add it to the libretro core. Yes. Its that simple. Love our code design :D

Cannot alter the window size though because the core has no access to the RetroArch Window.

```cpp
diff --git a/src/platform/libretro/ui.cpp b/src/platform/libretro/ui.cpp
index 602eaa1c..540cc2e9 100644
--- a/src/platform/libretro/ui.cpp
+++ b/src/platform/libretro/ui.cpp
@@ -57,6 +57,10 @@ Game_ConfigInput LibretroUi::cfg_input;
 static Input::Keys::InputKey RetroJKey2InputKey(int button_index);
 #endif
 
+// libretro needs an upper limit for the framebuffer set in av_info
+const int fb_max_width = 1920;
+const int fb_max_height = 1080;
+
 LibretroUi::LibretroUi(int width, int height, const Game_Config& cfg) : BaseUi(cfg)
 {
 	// Handled by libretro
@@ -79,7 +83,7 @@ LibretroUi::LibretroUi(int width, int height, const Game_Config& cfg) : BaseUi(c
 		PF::NoAlpha);
 
 	Bitmap::SetFormat(Bitmap::ChooseFormat(format));
-	main_surface.reset();
+
 	main_surface = Bitmap::Create(current_display_mode.width,
 		current_display_mode.height,
 		false,
@@ -103,6 +107,39 @@ void LibretroUi::UpdateDisplay() {
 	UpdateWindow(main_surface->pixels(), current_display_mode.width, current_display_mode.height, main_surface->pitch());
 }
 
+bool LibretroUi::ChangeDisplaySurfaceResolution(int new_width, int new_height) {
+	if (new_width == current_display_mode.width && new_height == current_display_mode.height) {
+		return true;
+	}
+
+	if (new_width > fb_max_width || new_height > fb_max_height) {
+		Output::Warning("ChangeDisplaySurfaceResolution: {}x{} is too large", new_width, new_height);
+		return false;
+	}
+
+	BitmapRef new_main_surface = Bitmap::Create(new_width, new_height, false, current_display_mode.bpp);
+
+	if (!new_main_surface) {
+		Output::Warning("ChangeDisplaySurfaceResolution Bitmap::Create failed");
+		return false;
+	}
+
+	retro_game_geometry geom = {};
+	geom.base_width = new_width;
+	geom.base_height = new_height;
+	if (!LibretroUi::environ_cb(RETRO_ENVIRONMENT_SET_GEOMETRY, &geom)) {
+		Output::Warning("ChangeDisplaySurfaceResolution SET_GEOMETRY failed");
+		return false;
+	}
+
+	main_surface = new_main_surface;
+
+	current_display_mode.width = new_width;
+	current_display_mode.height = new_height;
+
+	return true;
+}
+
 void LibretroUi::ProcessEvents() {
 #	if defined(USE_JOYSTICK) && defined(SUPPORT_JOYSTICK)
 	if (CheckInputState == nullptr) {
@@ -326,8 +363,8 @@ RETRO_API void retro_get_system_info(struct retro_system_info* info) {
 RETRO_API void retro_get_system_av_info(struct retro_system_av_info* info) {
 	info->geometry.base_width = SCREEN_TARGET_WIDTH;
 	info->geometry.base_height = SCREEN_TARGET_HEIGHT;
-	info->geometry.max_width = SCREEN_TARGET_WIDTH;
-	info->geometry.max_height = SCREEN_TARGET_HEIGHT;
+	info->geometry.max_width = fb_max_width;
+	info->geometry.max_height = fb_max_height;
 	info->geometry.aspect_ratio = 0.0f;
 	info->timing.fps = Game_Clock::GetTargetGameFps();
 	info->timing.sample_rate = AUDIO_SAMPLERATE;
diff --git a/src/platform/libretro/ui.h b/src/platform/libretro/ui.h
index a5a8418b..e9346662 100644
--- a/src/platform/libretro/ui.h
+++ b/src/platform/libretro/ui.h
@@ -45,6 +45,7 @@ public:
 	 * Inherited from BaseUi.
 	 */
 	/** @{ */
+	bool ChangeDisplaySurfaceResolution(int new_width, int new_height) override;
 	void UpdateDisplay() override;
 	void ProcessEvents() override;
 	void vGetConfig(Game_ConfigVideo& cfg) const override;
```

**Comment by BlisterB** [comment]
> 

It seems not to be possible to apply this patch:
"error: corrupt patch at line 87"

**Comment by BlisterB** [comment]
https://github.com/EasyRPG/Player/blob/8d508cf91a0bc0306288c43e20009f3433477891/src/plane.cpp#L50

The problem of panorama alignement seems to be linked with this ``ox`` variable that return the wrong value (ex. 14 instead of 16px).

File: src/game_map.cpp [hunk 1]
```diff
@@ -1730,8 +1730,9 @@ void Game_Map::Parallax::ResetPositionX() {
 	}
 
 	if (!params.scroll_horz && !LoopHorizontal()) {
-		if (GetWidth() > 20 && parallax_width > Player::screen_width) {
-			const int w = (GetWidth() - 20) * TILE_SIZE;
+		int nbTilesPerScreen = Player::screen_width / TILE_SIZE;
```
Comment (Ghabry) [inline src/game_map.cpp:1733]: Thanks for tracking down what is going wrong. This and your ``14`` gave me a hint to work with.

To get the correct value this division must be float and rounded up.

Examples:

```
400 / 16 = 25 <- ok
399 / 16 = 24 <- bad, must be 25
385 / 16 = 24 <- bad, must be 25
384 / 16 = 24 <- ok
```

Though this is inefficient. Thankfully this is a power of two so much faster bit trickery to the rescue:

```
int tiles_per_screen = ((Player::screen_width & ~0xF) + 16) / TILE_SIZE;
```

This AND-masks the lowest 4 Bit away, so 399 becomes 384 and then add 16 to reach 400.

**Comment by BlisterB** [comment]
Thank you Ghabry, it indeed solved the problem of fixed panorama that used to move.

However it now displays a space outside the map (at the right on this screenshot):
image

**Comment by BlisterB** [comment]
Correction: this visual bug was due to a screen width of 426, not a multiple of 16.

**Comment by BlisterB** [comment]
We will get a lot of feedbacks from this issue as 426x240 is the closest 16:9 ratio from RPG Maker 200X. 
I'll try to investigate on this subject.

**Comment by Ghabry** [comment]
You mean this artifact that happens when nothing is drawn for one frame? Take a look at "DrawBackground" of Scene_Map

**Review by Ghabry — APPROVED** [review]

**Review by fdelapena — APPROVED** [review]

**Comment by Ghabry** [comment]
Custom Resolution 3/? is on the way:

screenshot_0

Yes Jetrotal already asked for a customizable size dialog but lets start simple here until the major visual glitches are sorted out ;).

**Pull request merged**